packages/lesswrong/components/hooks/useBookmark.tsx

```diff
@@ -0,0 +1,112 @@
+import React, { useState, MouseEvent, useEffect, useCallback } from "react";
+import { gql, useMutation } from "@apollo/client";
+import { useCurrentUser } from "@/components/common/withUser";
+import { useDialog } from "@/components/common/withDialog";
+import { useTracking } from "@/lib/analyticsEvents";
+import type { ForumIconName } from "@/components/common/ForumIcon";
+import { Components } from "@/lib/vulcan-lib/components";
+import { fragmentTextForQuery } from "@/lib/vulcan-lib/fragments";
+import { useMulti } from "@/lib/crud/withMulti";
+
+export interface UseBookmarkResult {
+  isBookmarked: boolean;
+  toggleBookmark: (event?: MouseEvent) => void;
+  loading: boolean;
+  icon: ForumIconName;
+  labelText: string;
+  hoverText: string;
+}
+
+const TOGGLE_BOOKMARK_MUTATION = gql`
```

Comment:
These should not be defined at the top level of the module for gql parsing cost reasons; define it in the component function